lib/cloud_watch.ex

```diff
@@ -90,20 +110,76 @@ defmodule CloudWatch do
     %{state | buffer: [], buffer_length: 0, buffer_size: 0}
   end
 
-  defp add_message(%{buffer: buffer, buffer_length: buffer_length, buffer_size: buffer_size} = state, level, msg, ts, md) do
-    message = state.format
-              |> Logger.Formatter.format(level, msg, ts, md)
-              |> IO.chardata_to_string
-    #buffer = List.insert_at(buffer, -1, %InputLogEvent{message: message, timestamp: ts}) # performance impact of adding at the end?
-    buffer = [%InputLogEvent{message: message, timestamp: ts} | buffer] # buffer order is not relevant, we'll reverse or sort later if needed
-    %{state | buffer: buffer, buffer_length: buffer_length + 1, buffer_size: buffer_size + byte_size(message) + 26}
+  defp add_message(
+         %{buffer: buffer, buffer_length: buffer_length, buffer_size: buffer_size} = state,
+         level,
+         msg,
+         ts,
+         md
+       ) do
+    message =
+      level
+      |> format_event(msg, ts, md, state)
+      |> IO.chardata_to_string()
+
+    # buffer order is not relevant, we'll reverse or sort later if needed
+    buffer = [%InputLogEvent{message: message, timestamp: ts} | buffer]
+
+    %{
+      state
+      | buffer: buffer,
+        buffer_length: buffer_length + 1,
+        buffer_size: buffer_size + byte_size(message) + 26
+    }
+  end
+
+  @doc false
+  @spec metadata_matches?(Keyword.t(), nil | Keyword.t()) :: true | false
+  def metadata_matches?(_md, nil), do: true
+  # all of the filter keys are present
+  def metadata_matches?(_md, []), do: true
+
+  def metadata_matches?(md, [{key, val} | rest]) do
+    case Keyword.fetch(md, key) do
+      {:ok, ^val} ->
+        metadata_matches?(md, rest)
+
+      # fail on first mismatch
+      _ ->
+        false
+    end
+  end
+
+  defp take_metadata(metadata, :all), do: metadata
+
+  defp take_metadata(metadata, keys) do
+    metadatas =
```

Comment:
The assignment is not required here. You can just pipe the result straight into `Enum.reverse`. Maybe it would be nice as well to extract the reduce function to a separate function although I'm having trouble coming up with a good name for it now. I can live with the anonymous function as well.

Reply:
I just figured out that using `Keyword.take/2` here is simpler and more clear.